drivers/sensor/MLX90640.zig

```diff
@@ -0,0 +1,882 @@
+        ct: [5]i16 = undefined,
+        alpha: [768]u16 = undefined,
+        alphaScale: u8 = 0,
+        offset: [768]i32 = undefined,
+        kta: [768]i8 = undefined,
+        ktaScale: u8 = 0,
+        kv: [768]i8 = undefined,
+        kvScale: u8 = 0,
+        cpAlpha: [2]f32 = undefined,
+        cpOffset: [2]i16 = undefined,
+        ilChessC: [3]f32 = undefined,
+        brokenPixels: [5]u16 = undefined,
+        outlierPixels: [5]u16 = undefined,
+    };
+
+    eeprom: [832]u16 = undefined,
+    frame: [834]u16 = undefined,
+    frame_data: [834 * 2]u8 = undefined,
+    scratch_data: [768]f32 = undefined,
+    i2c: mdf.base.Datagram_Device,
+    clock_device: mdf.base.Clock_Device,
+    params: parameters,
+    emissivity: f32,
+    open_air_shift: f32,
+
+    scale_alpha: f32 = 0.000001,
+    control_register_address: u16 = 0x800D,
+    status_register_address: u16 = 0x8000,
+    ram_address: u16 = 0x0400,
+    eeprom_address: u16 = 0x2400,
+    device_id_1_address: u16 = 0x2407,
+
+    frame_loop: [2]u1 = [2]u1{ 0, 1 },
+    refresh_rate_mask: u16 = 0b111 << 7,
+
+    pub fn init(cfg: MLX90640_Config) !Self {
+        var obj = Self{
+            .i2c = cfg.i2c,
+            .clock_device = cfg.clock,
+            .open_air_shift = cfg.open_air_shift,
+            .emissivity = cfg.emissivity,
+            .params = parameters{},
+        };
+
+        // the temperatures are all wack on first read
+        try obj.loadFrame();
+
+        return obj;
+    }
+
+    fn write(self: *Self, address: u16, val: u48) !void {
+        var req = [4]u8{
+            @as(u8, @truncate(address >> 8)),
+            @as(u8, @truncate(address & 0xFF)),
+            @as(u8, @truncate(val >> 8)),
+            @as(u8, @truncate(val & 0xFF)),
+        };
+
+        try self.i2c.write(&req);
+    }
+
+    fn writeThenRead(self: *Self, address: u16, buf: []u16) !void {
+        var req = [2]u8{ @as(u8, @truncate(address >> 8)), @as(u8, @truncate(address & 0xFF)) };
+        try self.i2c.write_then_read(&req, self.frame_data[0 .. buf.len * 2]);
+        for (0.., buf) |i, _| {
+            buf[i] = std.mem.readInt(u16, self.frame_data[i * 2 .. (i * 2) + 2][0..2], .big);
+        }
+    }
+
+    pub fn serialNumber(self: *Self) !u48 {
+        try self.writeThenRead(self.device_id_1_address, self.frame[0..3]);
+        return @as(u48, self.frame[0]) << 32 |
+            @as(u48, self.frame[1]) << 16 |
+            @as(u48, self.frame[2]);
+    }
+
+    pub fn setRefreshRate(self: *Self, rate: u16) !void {
+        try self.writeThenRead(self.control_register_address, self.frame[0..1]);
+        const val: u16 = (self.frame[0] & ~self.refresh_rate_mask) | ((rate << 7) & self.refresh_rate_mask);
+        try self.write(self.control_register_address, val);
+    }
+
+    pub fn refreshRate(self: *Self) !u3 {
+        try self.writeThenRead(self.control_register_address, self.frame[0..1]);
+        const val = self.frame[0] >> 7 & 0b111;
```

Comment:
Could just define a type for this register read, bit cast it, then return the relevant field